atlasdb-jepsen-tests/src/main/java/com/palantir/atlasdb/jepsen/lock/LockCorrectnessChecker.java

```diff
@@ -0,0 +1,188 @@
+ *
+ * Licensed under the BSD-3 License (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://opensource.org/licenses/BSD-3-Clause
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.palantir.atlasdb.jepsen.lock;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Range;
+import com.google.common.collect.TreeRangeSet;
+import com.palantir.atlasdb.jepsen.CheckerResult;
+import com.palantir.atlasdb.jepsen.ImmutableCheckerResult;
+import com.palantir.atlasdb.jepsen.events.Checker;
+import com.palantir.atlasdb.jepsen.events.Event;
+import com.palantir.atlasdb.jepsen.events.EventVisitor;
+import com.palantir.atlasdb.jepsen.events.FailEvent;
+import com.palantir.atlasdb.jepsen.events.InfoEvent;
+import com.palantir.atlasdb.jepsen.events.InvokeEvent;
+import com.palantir.atlasdb.jepsen.events.OkEvent;
+import com.palantir.atlasdb.jepsen.events.RequestType;
+
+import com.palantir.util.Pair;
+
+/**
+ * Checker verifying that whenever a lock is granted, there was a time point between the request and the
+ * acknowledge when the lock was actually free to be granted. This is tricky due to the existence of refreshes
+ * and the uncertainty of the exact times because of the latency between requests and replies.
+ *
+ * A successful refresh means the lock was held in the (open) interval
+ * (a, b), where
+ * a is the OkEvent.time() of the last successful lock,
+ * b is the InvokeEvent.time() of the refresh,
+ * <i>>assuming</i> a successful refresh guarantees holding the lock since the last time the lock was granted. This will
+ * be verified by another checker.
+ *
+ * A successful unlock can be treated the same as a refresh, except that it has implications for further refreshes
+ * and unlocks (that we will check by other checkers).
+ *
+ * A successful lock means the lock was held <i>at some point</i> in the (closed) interval
+ * [a, b], where
+ * a is the InvokeEvent.time() of the lock request, and
+ * b is its OkEvent.time()
+ *
+ * We mantain a set L of all intervals where we know there was a lock, and a separate list U of all intervals where
+ * there was a lock at some point. In the latter list, we purposefully do not merge intervals, to ensure no information
+ * is lost.
+ *
+ * To verify locks were correctly granted, it is necessary and sufficient to verify that none of the uncertain
+ * intervals in U are fully covered by the intervals in the set of lock intervals L.
+ */
```

Comment:
Great job! 👍  At some point I think it would be nice to put something like this into the docs.